src/test/java/io/confluent/connect/elasticsearch/ElasticsearchClientTest.java

```diff
@@ -0,0 +1,526 @@
+    client.createIndex(INDEX);
+
+    writeRecord(sinkRecord("key0", 0), client);
+    writeRecord(sinkRecord("key1", 1), client);
+    client.flush();
+
+    waitUntilRecordsInES(2);
+
+    // delete 1
+    SinkRecord deleteRecord = new SinkRecord(INDEX, 0, Schema.STRING_SCHEMA, "key0", null, null, 3);
+    writeRecord(deleteRecord, client);
+
+    waitUntilRecordsInES(1);
+    client.close();
+  }
+
+  @Test
+  public void testUpsertRecord() throws Exception {
+    props.put(WRITE_METHOD_CONFIG, WriteMethod.UPSERT.name());
+    props.put(IGNORE_KEY_CONFIG, "false");
+    config = new ElasticsearchSinkConnectorConfig(props);
+    converter = new DataConverter(config);
+    ElasticsearchClient client = new ElasticsearchClient(config, null);
+    client.createIndex(INDEX);
+
+    writeRecord(sinkRecord("key0", 0), client);
+    writeRecord(sinkRecord("key1", 1), client);
+    client.flush();
+
+    waitUntilRecordsInES(2);
+
+    // create modified record for upsert
+    Schema schema = SchemaBuilder
+        .struct()
+        .name("record")
+        .field("offset", SchemaBuilder.int32().defaultValue(0).build())
+        .field("another", SchemaBuilder.int32().defaultValue(0).build())
+        .build();
+
+    Struct value = new Struct(schema).put("offset", 2);
+    SinkRecord upsertRecord = new SinkRecord(INDEX, 0, Schema.STRING_SCHEMA, "key0", schema, value, 2);
+
+    // upsert 1, write another
+    writeRecord(upsertRecord, client);
+    writeRecord(sinkRecord("key2", 3), client);
+    client.flush();
+
+    waitUntilRecordsInES(3);
+    for (SearchHit hit : helperClient.search(INDEX)) {
+      if (hit.getId().equals("key0")) {
+        assertEquals(2, hit.getSourceAsMap().get("offset"));
+        assertEquals(0, hit.getSourceAsMap().get("another"));
+      }
+    }
+
+    client.close();
+  }
+
+  @Test
+  public void testIgnoreBadRecord() throws Exception {
+    props.put(BEHAVIOR_ON_MALFORMED_DOCS_CONFIG, BehaviorOnMalformedDoc.IGNORE.name());
+    config = new ElasticsearchSinkConnectorConfig(props);
+    converter = new DataConverter(config);
+
+    ElasticsearchClient client = new ElasticsearchClient(config, null);
+    client.createIndex(INDEX);
+    client.createMapping(INDEX, schema());
+
+    Schema schema = SchemaBuilder
+        .struct()
+        .name("record")
+        .field("not_mapped_field", SchemaBuilder.int32().defaultValue(0).build())
+        .build();
+    Struct value = new Struct(schema).put("not_mapped_field", 420);
+    SinkRecord badRecord = new SinkRecord(INDEX, 0, Schema.STRING_SCHEMA, "key", schema, value, 0);
+
+    writeRecord(sinkRecord(0), client);
+    client.flush();
+
+    writeRecord(badRecord, client);
+    client.flush();
+
+    writeRecord(sinkRecord(1), client);
+    client.flush();
+
+    waitUntilRecordsInES(2);
+    assertEquals(2, helperClient.getDocCount(INDEX));
+    client.close();
+  }
+
+  @Test(expected = ConnectException.class)
```

Comment:
It'd be nice to start using assertThrows and available hamcrest matchers for maps etc. But we can consider on a follow up.